File: stealth/semaphore/waitfor.f
template <typename Rep, typename Period>
std::cv_status Semaphore::wait_for(
                        std::chrono::duration<Rep, Period> const &relTime)
{
    std::unique_lock<std::mutex> lk(d_mutex);   // get the lock
    while (d_nAvailable == 0)
        if (d_condition.wait_for(lk, relTime) == std::cv_status::timeout)
            return std::cv_status::timeout;

    --d_nAvailable;
    return std::cv_status::no_timeout;
}

File: stealth/wait11.OBS/waitfor.cc
#include "wait11.ih"

void Wait11::waitFor(size_t seconds, bool doM2)
{
    d_signaled = false;

    while (true)
    {
        unique_lock<mutex> lock(d_mutex);    // get the lock

        if (d_condition.wait_for(lock, chrono::seconds(seconds)) ==
                            std::cv_status::timeout)
        {
            if (doM2)
                m2 << "timeout after " << seconds << " seconds" << endl;
            return;
        }

        if (d_signaled)
        {
            if (doM2)
                m2 << "received signal: done waiting" << endl;
            return;
        }
    }
}
